concordium-consensus/tests/consensus/ConcordiumTests/KonsensusV1/Types.hs

```diff
@@ -0,0 +1,174 @@
+    tmEpochFinalizationEntry <- oneof [return Absent, Present <$> genFinalizationEntry]
+    tmAggregateSignature <- TimeoutSignature <$> genBlsSignature
+    return TimeoutMessageBody{..}
+
+-- |Generate a 'TimeoutMessage' signed by an arbitrarily-generated keypair.
+genTimeoutMessage :: Gen TimeoutMessage
+genTimeoutMessage = do
+    body <- genTimeoutMessageBody
+    kp <- genBlockKeyPair
+    return $ signTimeoutMessage body kp
+
+-- |Check that serialization followed by deserialization gives the identity.
+serCheck :: (Eq a, Serialize a, Show a) => a -> Property
+serCheck a = decode (encode a) === Right a
+
+-- |Test that serializing then deserializing a finalizer set is the identity.
+propSerializeFinalizerSet :: Property
+propSerializeFinalizerSet = forAll genFinalizerSet serCheck
+
+-- |Test that serializing then deserializing a quorum certificate is the identity.
+propSerializeQuorumCertificate :: Property
+propSerializeQuorumCertificate = forAll genQuorumCertificate serCheck
+
+-- |Test that serializing then deserializing a finalization entry is the identity.
+propSerializeFinalizationEntry :: Property
+propSerializeFinalizationEntry = forAll genFinalizationEntry serCheck
+
+-- |Test that serializing then deserializing a timeout certificate is the identity.
+propSerializeTimeoutCertificate :: Property
+propSerializeTimeoutCertificate = forAll genTimeoutCertificate serCheck
+
+-- |Test that serializing then deserializing a timeout message body is the identity.
+propSerializeTimeoutMessageBody :: Property
+propSerializeTimeoutMessageBody = forAll genTimeoutMessageBody serCheck
+
+-- |Test that serializing then deserializing a timeout message is the identity.
+propSerializeTimeoutMessage :: Property
+propSerializeTimeoutMessage = forAll genTimeoutMessage serCheck
+
+-- |Check that a signing a timeout message produces a timeout message that verifies with the key.
+propSignTimeoutMessagePositive :: Property
+propSignTimeoutMessagePositive =
+    forAll genTimeoutMessageBody $ \body ->
+        forAll genBlockKeyPair $ \kp ->
+            checkTimeoutMessageSignature (Sig.verifyKey kp) (signTimeoutMessage body kp)
+
+-- |Check that a signing a timeout message produces a timeout message that does not verify with a
+-- different key.
+propSignTimeoutMessageDiffKey :: Property
+propSignTimeoutMessageDiffKey =
+    forAll genTimeoutMessageBody $ \body ->
+        forAll genBlockKeyPair $ \kp1 ->
+            forAll genBlockKeyPair $ \kp2 ->
+                (kp1 /= kp2) ==>
+                    not (checkTimeoutMessageSignature (Sig.verifyKey kp2) (signTimeoutMessage body kp1))
+
+-- |Check that signing a timeout message and changing the body to something different produces a
+-- timeout message that does not verify with the key.
+propSignTimeoutMessageDiffBody :: Property
+propSignTimeoutMessageDiffBody =
+    forAll genTimeoutMessageBody $ \body1 ->
+        forAll genTimeoutMessageBody $ \body2 ->
+            (body1 /= body2) ==>
```

Comment:
That's probably not a very efficient test. It'd probably be better to test that flipping bits in the serialized message makes it fail either parsing of signature checking.